test/extended/apiserver/latency.go

```diff
@@ -0,0 +1,101 @@
+)
+
+// TODO: remove these variables once we move to route instead of exec pod
+var (
+	oc                   *exutil.CLI
+	url, bearerToken, ns string
+	execPod              *v1.Pod
+)
+
+var _ = g.Describe("[sig-instrumentation][Late] API Server latency", func() {
+	defer g.GinkgoRecover()
+
+	oc = exutil.NewCLIWithoutNamespace("apiserver-latency")
+
+	g.BeforeEach(func() {
+		err := exutil.WaitForAnImageStream(
+			oc.AdminImageClient().ImageV1().ImageStreams("openshift"), "tools",
+			exutil.CheckImageStreamLatestTagPopulated, exutil.CheckImageStreamTagNotFound,
+		)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		var ok bool
+		url, _, bearerToken, ok = promhelper.LocatePrometheus(oc)
+		if !ok {
+			e2e.Failf("Prometheus could not be located on this cluster, failing prometheus test")
+		}
+	})
+
+	// Lantecy thresholds collected from:
+	// - https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.9-e2e-gcp-upgrade/1480678513441771520
+	// - https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.9-e2e-gcp/1479392960888967168
+	// - https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.9-e2e-aws-serial/1479392960926715904
+	g.It("should verify that the apiserver requests latency are within expected thresholds", func() {
+		ns = oc.SetupNamespace()
+
+		// TODO: access Prometheus via route instead of exec pod
+		execPod = exutil.CreateExecPodOrFail(oc.AdminKubeClient(), ns, "execpod")
+		defer func() {
+			oc.AdminKubeClient().CoreV1().Pods(ns).Delete(context.Background(), execPod.Name, *metav1.NewDeleteOptions(1))
+		}()
+
+		// 99th percentile of resource-scoped requests latency
+		scope := "resource|"
+		expectRequestLatency("GET", scope, parseDuration("75ms"))
+		expectRequestLatency("POST", scope, parseDuration("200ms"))
+		expectRequestLatency("PUT", scope, parseDuration("75ms"))
+		expectRequestLatency("PATCH", scope, parseDuration("75ms"))
+		expectRequestLatency("DELETE", scope, parseDuration("75ms"))
+
+		// 99th percentile of namespace-scoped requests latency
+		scope = "namespace"
+		expectRequestLatency("LIST", scope, parseDuration("250ms"))
+		expectRequestLatency("POST", scope, parseDuration("750ms"))
+		expectRequestLatency("DELETE", scope, parseDuration("750ms"))
+
+		// 99th percentile of cluster-scoped requests latency
+		scope = "cluster"
+		expectRequestLatency("LIST", scope, parseDuration("1s"))
+		expectRequestLatency("POST", scope, parseDuration("500ms"))
+	})
+})
+
+func parseDuration(s string) time.Duration {
+	d, err := time.ParseDuration(s)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	return d
+}
+
+func expectRequestLatency(verb string, scope string, threshold time.Duration) {
+	resp, err := promhelper.RunQuery(
+		fmt.Sprintf(
+			`histogram_quantile(0.99, sum(rate(apiserver_request_duration_seconds_bucket{job="apiserver",verb="GET",scope=~"%s"}[%s])) by (le))`,
```

Comment:
why is this hardcoded to GET instead of substituting "verb"

Reply:
:facepalm: thank you for catching that :laughing: